File: BinaryTreeRightSide.cpp
//Time Complexity: O(n)
//Space complexity: O(n)

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> result; 
    vector<int> rightSideView(TreeNode* root) {
        if(root == nullptr) return {}; 
        bfs(root); 
        return result;     
    }

    void bfs(TreeNode* node) {
        deque<TreeNode*> q; 
        q.push_back(node); 

        while(!q.empty()) {
            int size = q.size(); 
            while(size > 1) {
                TreeNode* n = q.front();
                q.pop_front();  
                if(n->left != nullptr) q.push_back(n->left); 
                if(n->right != nullptr) q.push_back(n->right); 
                size--; 
            }
            TreeNode* n = q.front(); 
            q.pop_front();
            result.push_back(n->val); 
            if(n->left != nullptr) q.push_back(n->left); 
            if(n->right != nullptr) q.push_back(n->right);  
        }
    }
};

File: binaryTreeRightSide.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# creating a queue of treeNode objects
# define a rightSide node to pick up elements level by level
# then run a for loop in each level , pop the left element in the queue
# add the children babies to the queue
# append the level result to res
# time complexity 0(n) because you visit all the nodes 
# space complexity 0(n) because queue to store nodes 


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        q = collections.deque([root])

        while q:
            rightSide = None
            qLen = len(q)

            for i in range(qLen):
                node = q.popleft()
                if node :
                    rightSide = node
                    q.append(node.left)
                    q.append(node.right)

            if rightSide :
                res.append(rightSide.val)
        
        return res